sopel/modules/emailcheck.py

```diff
@@ -0,0 +1,314 @@
+# coding=utf-8
+"""
+emailcheck.py - Watch oper messages for new nicks being registered
+Copyright © 2021, Kufat <kufat@kufat.net>
+Based on existing sopel code.
+Licensed under the Eiffel Forum License 2.
+"""
+
+import logging
+import re
+import urllib
+
+import sqlalchemy.sql
+
+from dataclasses import dataclass
+
+from sopel import db, module
+from sopel.config.types import FilenameAttribute, StaticSection, ValidatedAttribute, ListAttribute
+from sopel.tools import events, target, Identifier
+
+from sqlalchemy import Column, String, Float, Boolean, TIMESTAMP
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.ext.declarative import declarative_base
+
+try:
+    from ip import get_exemption
+    from ip import ipqs_lock
+except:
+    import threading
+
+    def get_exemption(ip):
+        return "Can't access exemptions; failing safe"
+
+    ipqs_lock = threading.Lock()
+
+
+EMAIL_REGEX = re.compile(r"([a-zA-Z0-9_.+-]+)@([a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+)")
+IRCCLOUD_USER_REGEX = re.compile(r"[us]id[0-9]{4,}")
+DOMAIN_LEN = 50
+
+DEFAULT_EXEMPT_SUFFIXES = {
+    "@gmail.com",
+    "@hotmail.com",
+    "@protonmail.com",
+    ".edu"
+}
+
+KILL_STR = ":Use of disposable email service for nick registration"
+
+LOGGER = logging.getLogger(__name__)
+
+BASE = declarative_base()
+
+safe_mode = True
+db_populated = False
+
+#SQLAlchemy container class
+class KnownEmails(BASE):
+    __tablename__ = 'known_emails'
+    domain = Column(String(DOMAIN_LEN), primary_key=True)
+    first_nick = Column(String(40))
+    score = Column(Float)
+    flag_disposable = Column(Boolean)
+    flag_recent_abuse = Column(Boolean)
+    first_seen = Column(TIMESTAMP, server_default=sqlalchemy.sql.func.now())
+
+class EmailCheckSection(StaticSection):
+    IPQS_key = ValidatedAttribute('IPQS_key')
+    disallow_threshold = ValidatedAttribute("disallow_threshold", parse=float, default=50.0)
+    malicious_threshold = ValidatedAttribute("malicious_threshold", parse=float, default=75.0)
+    gline_time = ValidatedAttribute('gline_time', default="24h")
+    exempt_suffixes = ListAttribute("exempt_suffixes")
+    warn_chans = ListAttribute("warn_chans")
+    protect_chans = ListAttribute("protect_chans")
```

Comment:
Nitpick: inconsistent strings between single and double quotes

Reply:
Will fix